contrib/format-daffodil/src/main/java/org/apache/drill/exec/store/daffodil/schema/DaffodilDataProcessorFactory.java

```diff
@@ -0,0 +1,165 @@
+ * obtain a DataProcessor for use with DaffodilMessageParser.
+ * <p/>
+ * TODO: Needs to use a cache to avoid reloading/recompiling every time.
+ */
+public class DaffodilDataProcessorFactory {
+  // Default constructor is used.
+
+  private static final Logger logger = LoggerFactory.getLogger(DaffodilDataProcessorFactory.class);
+
+  private DataProcessor dp;
+
+  /**
+   * Gets a Daffodil DataProcessor given the necessary arguments to compile or reload it.
+   *
+   * @param schemaFileURI
+   *     pre-compiled dfdl schema (.bin extension) or DFDL schema source (.xsd extension)
+   * @param validationMode
+   *     Use true to request Daffodil built-in 'limited' validation. Use false for no validation.
+   * @param rootName
+   *     Local name of root element of the message. Can be null to use the first element declaration
+   *     of the primary schema file. Ignored if reloading a pre-compiled schema.
+   * @param rootNS
+   *     Namespace URI as a string. Can be null to use the target namespace of the primary schema
+   *     file or if it is unambiguous what element is the rootName. Ignored if reloading a
+   *     pre-compiled schema.
+   * @return the DataProcessor
+   * @throws CompileFailure
+   *     - if schema compilation fails
+   */
+  public DataProcessor getDataProcessor(URI schemaFileURI, boolean validationMode, String rootName,
+      String rootNS)
+      throws CompileFailure {
+
+    DaffodilDataProcessorFactory dmp = new DaffodilDataProcessorFactory();
+    boolean isPrecompiled = schemaFileURI.toString().endsWith(".bin");
+    if (isPrecompiled) {
+      if (Objects.nonNull(rootName) && !rootName.isEmpty()) {
+        // A usage error. You shouldn't supply the name and optionally namespace if loading
+        // precompiled schema because those are built into it. Should be null or "".
+        logger.warn("Root element name '{}' is ignored when used with precompiled DFDL schema.",
+            rootName);
+      }
+      try {
+        dmp.loadSchema(schemaFileURI);
+      } catch (IOException | InvalidParserException e) {
+        throw new CompileFailure(e);
+      }
+      dmp.setupDP(validationMode, null);
+    } else {
+      List<Diagnostic> pfDiags;
+      try {
+        pfDiags = dmp.compileSchema(schemaFileURI, rootName, rootNS);
+      } catch (URISyntaxException | IOException e) {
+        throw new CompileFailure(e);
+      }
+      dmp.setupDP(validationMode, pfDiags);
+    }
+    return dmp.dp;
+  }
+
+  private void loadSchema(URI schemaFileURI) throws IOException, InvalidParserException {
+    Compiler c = Daffodil.compiler();
+    dp = c.reload(Channels.newChannel(schemaFileURI.toURL().openStream()));
+  }
+
+  private List<Diagnostic> compileSchema(URI schemaFileURI, String rootName, String rootNS)
+      throws URISyntaxException, IOException, CompileFailure {
+    Compiler c = Daffodil.compiler();
+    ProcessorFactory pf = c.compileSource(schemaFileURI, rootName, rootNS);
```

Comment:
Schema source files can be spread over multiple jars (sometimes dozens of files), and they contain relative paths within them that must resolve relative to that same classpath.

Reply:
Note: This compilation really wants to happen once on a central node and the resulting binary schema file wants to be cached somewhere that gets copied out to the drill bit locations. 

Right now this either loads a binary DFDL schema, which needs to happen on the drill bits, or it needs to compile a DFDL schema source (which needs to happen centrally) to create the binary DFDL schema, which then gets loaded by the drill bits. 

This has to get refactored because the drill bits should *never* be compiling a DFDL schema. They should only be loading up pre-compiled binary schemas.